File: python/test_binary_search.py
import pytest
import binary_search as bs

@pytest.fixture
def ary():
    return [1, 2, 3, 4]

def test_lower_bound(ary):
    assert ary[bs.lower_bound(ary, 3)] == 3
    assert ary[bs.lower_bound(ary, 2.5)] == 3
    assert ary[bs.lower_bound(ary, 1)] == 1
    assert bs.lower_bound(ary, 5) == len(ary)

def test_upper_bound(ary):
    assert bs.upper_bound(ary, 4) == len(ary)
    assert ary[bs.upper_bound(ary, 2.5)] == 3
    assert ary[bs.upper_bound(ary, 3)] == 4

def test_binary_search(ary):
    assert bs.binary_search(ary, 3)
    assert not bs.binary_search(ary, 5)

# the return value of equal_range is index instead of value
def test_equal_range(ary):
    assert bs.equal_range(ary, 3) == (2, 3)
    assert bs.equal_range(ary, 2.5) == (2, 2)

File: ruby/test_binary_search.rb
require 'minitest/autorun'
require_relative 'binary_search'

class TestBinarySearch < MiniTest::Test
  include BinarySearch

  def setup
    @a = [1, 2, 3, 4]
  end

  def test_lower_bound
    assert_equal 3, @a[lower_bound(@a, 3)]
    assert_equal 3, @a[lower_bound(@a, 2.5)]
    assert_equal 1, @a[lower_bound(@a, 1)]
    assert_equal @a.size, lower_bound(@a, 5)
  end

  def test_upper_bound
    assert_equal @a.size, upper_bound(@a, 4)
    assert_equal 3, @a[upper_bound(@a, 2.5)]
    assert_equal 4, @a[upper_bound(@a, 3)]
  end

  def test_binary_search
    assert_equal true, binary_search(@a, 3)
    assert_equal false, binary_search(@a, 0)
    assert_equal false, binary_search(@a, 5)
  end

  def test_equal_range
    assert_equal [2, 3], equal_range(@a, 3)
    assert_equal [2, 2], equal_range(@a, 2.5)
  end
end
